File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
// space:o(n)
// time:o(n)

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        // pair: [index, height]
        stack<pair<int, int>> stk;
        int result = 0;
        
        for (int i = 0; i < heights.size(); i++) {
            int start = i;
            
            while (!stk.empty() && stk.top().second > heights[i]) {
                int index = stk.top().first;
                int width = i - index;
                int height = stk.top().second;
                stk.pop();
                
                result = max(result, height * width);
                start = index;
            }
            
            stk.push({start, heights[i]});
        }
        
        while (!stk.empty()) {
            int width = heights.size() - stk.top().first;
            int height = stk.top().second;
            stk.pop();
            
            result = max(result, height * width);
        }
                          
        return result;
    }
};

/*
int main() {
    Solution solution;

    // Define the input values (list of heights)
    vector<int> heights = {2, 1, 5, 6, 2, 3};  // Example input

    // Call the largestRectangleArea function with the provided input
    int result = solution.largestRectangleArea(heights);

    // Print the result
    cout << "Largest rectangle area: " << result << endl;

    return 0;
}
*/

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
## time: o(n)
## space : o(n)

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        
        ## initialize maxarea as '0'
        maxArea = 0
        
        ## a stack with pair : (index:height)
        stack = []
        
        ## interate for both index and height in input array
        
        for i, h in enumerate(heights):
            
            ## let the start be i(initially)
            
            start = i
            
            ## when stack not empty and stack top value is greater than height
            
            while stack and stack[-1][1] > h:
                ## pop and store in index, height
                index, height = stack.pop()
                ## area is height * width where width is i-index
                maxArea = max(maxArea, height * (i-index))
                ##update start
                start = index
                
            stack.append((start, h))
            
        ## now check for the values left in the stack
        for i, h in stack:
            
            maxArea = max(maxArea, h * (len(heights) - i))
            
        return maxArea   
        
                
        